concordium-smart-contract-testing/src/impls.rs

```diff
@@ -0,0 +1,962 @@
+
+impl ChainEvent {
+    /// Get the contract address that this event relates to.
+    /// This means the `address` field for all variant except `Transferred`,
+    /// where it returns the `from`.
+    pub fn contract_address(&self) -> ContractAddress {
+        match self {
+            ChainEvent::Interrupted { address, .. } => *address,
+            ChainEvent::Resumed { address, .. } => *address,
+            ChainEvent::Upgraded { address, .. } => *address,
+            ChainEvent::Updated { address, .. } => *address,
+            ChainEvent::Transferred { from, .. } => *from,
+        }
+    }
+}
+
+/// Convert [`Energy`] to [`InterpreterEnergy`] by multiplying by `1000`.
+pub(crate) fn to_interpreter_energy(energy: Energy) -> InterpreterEnergy {
+    InterpreterEnergy::from(energy.energy * 1000)
+}
+
+/// Convert [`InterpreterEnergy`] to [`Energy`] by dividing by `1000`.
+pub(crate) fn from_interpreter_energy(interpreter_energy: InterpreterEnergy) -> Energy {
+    Energy::from(interpreter_energy.energy / 1000)
+}
+
+/// Calculate the energy energy for looking up a [`ContractModule`].
+pub(crate) fn lookup_module_cost(module: &ContractModule) -> Energy {
+    // The ratio is from Concordium/Cost.hs::lookupModule
+    Energy::from(module.size / 50)
+}
+
+/// Calculate the microCCD(mCCD) cost of energy(NRG) using the two exchange
+/// rates provided.
+// TODO: Find a way to make this parse the doc tests
+// To find the mCCD/NRG exchange rate:
+//
+//  euro     mCCD   euro * mCCD   mCCD
+//  ----  *  ---- = ----------- = ----
+//  NRG      euro   NRG * euro    NRG
+//
+// To convert the `energy` parameter to mCCD:
+//
+//        mCCD   NRG * mCCD
+//  NRG * ---- = ---------- = mCCD
+//        NRG       NRG
+pub(crate) fn energy_to_amount(
+    energy: Energy,
+    euro_per_energy: ExchangeRate,
+    micro_ccd_per_euro: ExchangeRate,
+) -> Amount {
+    let micro_ccd_per_energy_numerator: BigUint =
+        BigUint::from(euro_per_energy.numerator()) * micro_ccd_per_euro.numerator();
+    let micro_ccd_per_energy_denominator: BigUint =
+        BigUint::from(euro_per_energy.denominator()) * micro_ccd_per_euro.denominator();
+    let cost: BigUint =
+        (micro_ccd_per_energy_numerator * energy.energy) / micro_ccd_per_energy_denominator;
+    let cost: u64 = u64::try_from(cost).expect("Should never overflow due to use of BigUint");
+    Amount::from_micro_ccd(cost)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const ACC_0: AccountAddress = AccountAddress([0; 32]);
+    const ACC_1: AccountAddress = AccountAddress([1; 32]);
+
+    #[test]
+    fn calculate_cost_will_not_overflow() {
+        let micro_ccd_per_euro = ExchangeRate::new_unchecked(u64::MAX, u64::MAX - 1);
+        let euro_per_energy = ExchangeRate::new_unchecked(u64::MAX - 2, u64::MAX - 3);
+        let energy = Energy::from(u64::MAX - 4);
+        energy_to_amount(energy, euro_per_energy, micro_ccd_per_euro);
+    }
+
+    #[test]
+    fn creating_accounts_work() {
```

Comment:
I'll have a look at these tests again. I just added them during development to ensure that various things worked as expected. I might remove some of them again since they are redundant.